src/azul/deployment.py

```diff
@@ -427,8 +429,39 @@ def client(self, *args, **kwargs):
         Caching the result of this function is not necessary and will be harmful
         if the cached value is used by a thread other than the one that called
         this function.
+
+        Enable azul_logging in order to log Boto3 client's requests and
+        responses.
+        """
+        client = self.boto3_session.client(*args, **kwargs)
+        if azul_logging:
+            client.meta.events.register("before-send", self._log_client_request)
+            client.meta.events.register("response-received", self._log_client_response)
+        return client
+
+    def _log_client_request(self, event_name: str, request: Any, **kwargs) -> Optional[AWSResponse]:
+        """
+        Log the client's AWS API action, method, url being requested and the
+        body (if available). Logging of streamable body objects is skipped.
+        """
+        event_name = event_name.replace('before-send', 'Boto3 client')
+        log.info('%s requesting %s to %s', event_name, request.method, request.url)
+        if type(request.body) in (str, bytes, bytearray):
+            log.info('… with request body %r', trunc_ellipses(request.body, 128))
+        return None
+
+    def _log_client_response(self, *, event_name: str, **kwargs: Any) -> Optional[AWSResponse]:
+        """
+        This logs the response status of the client AWS API action and content
+        body (if available). Logging of streamable body objects is skipped.
         """
-        return self.boto3_session.client(*args, **kwargs)
+        event_name = event_name.replace('response-received', 'Boto3 client')
+        response = kwargs.get('response_dict')
+        body = response.get('body', '')
+        if type(body) in (str, bytes, bytearray):
```

Comment:
Would `body is not None` also work here? Or can there be other types besides str/bytes/bytearray?

Reply:
I don't think so. Some boto3 clients can stream data, like some s3 multipart operations, which I assume would be of a different type. For that reason, if memory serves, when I was drafting this with Hannes, he instructed me to conditionally log the `body` based on its type being one of str/bytes/bytearray.